File: NumberOfIslands/numberOfIslands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
  let numberOfIsland = 0;
  for (let row = 0; row < grid.length; row++) {
    
    for (let col = 0; col < grid[0].length; col++) {
      if (grid[row][col] == 1) {
        findAllLand(grid, row, col);
        numberOfIsland++;
      }
    }
  }
  return numberOfIsland;
};

function findAllLand(grid, row, col) {
  grid[row][col] = 0;

  // look right 
  if (col < grid[0].length - 1) {
    if (grid[row][col + 1] == 1) {
      findAllLand(grid, row, col + 1);
    }
  }

  // look down
  if (row < grid.length - 1) {
    if (grid[row + 1][col] == 1) {
      findAllLand(grid, row + 1, col)
    }
  }

  // look up
  if (row > 0) {
    if (grid[row - 1][col] == 1) {
      findAllLand(grid, row - 1, col)
    }
  }

  // look left
  if (col > 0) {
    if (grid[row][col - 1] == 1) {
      findAllLand(grid, row, col - 1)
    }
  }

  return;

}


const grid = [
  ["0","1","1","1","0"],
  ["1","0","0","0","0"],
  ["1","1","1","1","0"],
  ["0","0","0","0","1"]
];

numIslands(grid);

File: NumberOfIslands/numberOfIslands.py
class Solution:
    def numIslands(self, grid):
        if len(grid) == 0:
            return 0
        
        row = len(grid)
        col = len(grid[0])
        numberIslands = 0

        for r in range(row):
            for c in range(col):
                if grid[r][c] == "1":
                    numberIslands += 1
                    self.exploreIsland(grid, r, c)
        return numberIslands
    
    
    def exploreIsland(self, grid, row, col):
        if col < 0 or row < 0 or col >= len(grid[0]) or row >= len(grid) or grid[row][col] == '0':
            return
        
        grid[row][col] = "0"
        
        self.exploreIsland(grid, row, col + 1)
        self.exploreIsland(grid, row + 1, col)
        self.exploreIsland(grid, row, col - 1)
        self.exploreIsland(grid, row - 1, col)

        return grid
    
grid = [
  ["1","1","1","1","0"],
  ["1","1","0","1","1"],
  ["1","1","0","0","0"],
  ["0","0","1","0","1"]
]
sol = Solution()
print(sol.numIslands(grid))
